pkg/xds/experimental/client/parse.go

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package client
+
+import (
+	"fmt"
+
+	clusterpb "github.com/cilium/proxy/go/envoy/config/cluster/v3"
+	endpointpb "github.com/cilium/proxy/go/envoy/config/endpoint/v3"
+	listenerpb "github.com/cilium/proxy/go/envoy/config/listener/v3"
+	routepb "github.com/cilium/proxy/go/envoy/config/route/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/cilium/cilium/pkg/envoy"
+)
+
+func parseResource(typeUrl string, res *anypb.Any) (proto.Message, string, error) {
```

Comment:
consider just using the type directly, this may avoid having to pass around the typeURL

```
func parseResource(...) {

}

func resourceName(res *anypb.Any) (string, error) {
  var name string
  switch obj := res.(type) {
  case listenerpb.Listener:
    name = obj.Name
  ...
  default:
    return "", fmt.Errorf("invalid msg type: %T", msg)
  } 
  return name, nil
}
```

Reply:
We need to check that the parsed resource matches declared `typeUrl`. With your proposal we would accept:
```
DiscoveryResponse{
  typeUrl: Endpoint,
  resources: Any{
    typeUrl: Endpoint,
    res: Listener
  }
}
```
Fixing it would result in very similar code.

Reply:
It seems that due to the way `UnmarshalNew` works, it will always return resource of a `type` corresponding to `typeUrl`, so this check is redundant. I applied your suggestion and streamlined the code even further. Thank you